### PR TITLE
Update internal code for torch.linalg.solve

### PR DESCRIPTION
**Pull request by IvanYashchuk**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56613 Update internal code for torch.linalg.solve**

Replace linalg_solve_helper with `lu_stub` + `lu_solve_stub`.
Once `lu_stub` and `lu_solve_stub` have cuSOLVER-based codepath,
`torch.linalg.solve` will have it as well.

Differential Revision: [D28627408](https://our.internmc.facebook.com/intern/diff/D28627408)

### CODE REVIEW

File: test/test_linalg.py [hunk 1]
```diff
@@ -3255,6 +3255,7 @@ def solve_test_helper(self, A_dims, b_dims, device, dtype):
         A = random_fullrank_matrix_distinct_singular_value(*A_dims, dtype=dtype).to(device)
         return b, A
 
+    @onlyOnCPUAndCUDA
```
Comment (IvanYashchuk) [inline test/test_linalg.py:3258]: Skips meta device because of `RuntimeError: DispatchStub: unsupported device typemeta`.

**Comment by mruberry** [comment]
This entire stack looks good to me, @nikitaved would you take a closer look just to sanity check?

**Comment by IvanYashchuk** [comment]
I need to investigate the Windows build failure.

**Review by nikitaved — COMMENTED** [review]
Apologies for the delay! LGTM!

**Review by mruberry — APPROVED** [review]
Stamped! Thanks @IvanYashchuk and @nikitaved!

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28248766).

**Comment by IvanYashchuk** [comment]
Note: tests pass with `CUDA_LAUNCH_BLOCKING=1` in the above failing CI builds.

**Comment by IvanYashchuk** [comment]
@mruberry, I've submitted the changes from the `ci-all` branch, here is the draft PR https://github.com/pytorch/pytorch/pull/57993 for the CI status; it seems that the problem with Windows builds is gone.

The solution was to add a barrier after the call to the MAGMA routine for the batched LU decomposition that prevents premature usage of the data in the next MAGMA calls.

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28379394).

File: test/test_linalg.py (hunk 1 above)
Comment (walterddr) [reply test/test_linalg.py:3398]: Seems like there are additional tests that needs to be skipped. however I can't be sure. See: 
https://app.circleci.com/pipelines/github/pytorch/pytorch/322056/workflows/fe2c23e9-fe0a-4d1f-8b55-2877c77c48eb/jobs/13459579

**Review by walterddr — COMMENTED** [review]
reverting this PR as it broke no-arch test on master.

**Comment by IvanYashchuk** [comment]
@mruberry, the problem should be fixed now. The issue was that I forgot to update the dispatch key of `linalg_solve`. Since the implementation of this function now uses DispatchStub, we need to have `CPU, CUDA` dispatch key instead of `CompositeExplicitAutograd`.

**Comment by walterddr** [comment]
not exactly sure what went wrong in previous round of ci-all tests in #57993 but it did show noarch tests passing 
(in fact those failed master tests were skipped on your ci-all PR)
See:
https://app.circleci.com/pipelines/github/pytorch/pytorch/322056/workflows/fe2c23e9-fe0a-4d1f-8b55-2877c77c48eb/jobs/13459579
vs.
https://app.circleci.com/pipelines/github/pytorch/pytorch/319696/workflows/34bd1c85-0d8d-4db5-b0f0-056b9fba17e8/jobs/13340149

I will try to dig deeper into the reason behind this

**Comment by IvanYashchuk** [comment]
@walterddr,  `at::tensorsolve` started to use `at::linalg_solve` that is being modified in this PR instead of the deprecated `at::solve`. The PR for `at::tensorsolve` was merged 2 days ago https://github.com/pytorch/pytorch/pull/57741.
So it's just two related PRs merged simultaneously without being tested against each other. Now it should be safe to merge this PR, ci-all tests in #57993 pass except one failure that seems to be unrelated.

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28627408).

**Comment by mruberry** [comment]
@IvanYashchuk Sorry I missed the update on this. I've reimported it so we can land it. Just ping me if this is still good to go!

**Comment by IvanYashchuk** [comment]
@mruberry, could you please import this again? I rebased this PR, let's get it merged.

The draft PR #57993 is for the CI-all status.

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28627408).

**Pull request closed without merging**